### PR TITLE
Feat/ Custom hook [#224]

### PR DESCRIPTION
**Pull request by leomp12**
😛 

fixes https://github.com/fastify/fastify-rate-limit/issues/224

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/global-rate-limit.test.js [hunk 1]
```diff
@@ -293,6 +299,11 @@ test('With redis store', async t => {
 })
 
 test('Skip on redis error', async t => {
+  if (isSkipRedis) {
+    t.pass('Redis is not available')
+    return
+  }
```
Comment (mcollina) [inline test/global-rate-limit.test.js:305]: test have a `skip` option, use that instead

File: test/global-rate-limit.test.js [hunk 2]
```diff
@@ -1002,7 +989,7 @@ test('When continue exceeding is on (Local)', async t => {
 test('When continue exceeding is on (Redis)', async t => {
   const fastify = Fastify()
 
-  const redis = new Redis({ host: REDIS_HOST })
+  const redis = isSkipRedis ? undefined : new Redis({ host: REDIS_HOST })
```
Comment (mcollina) [inline test/global-rate-limit.test.js:992]: test have a `skip` option, use that instead

File: test/types/types.test.ts [hunk 3]
```diff
@@ -35,7 +35,7 @@ const options1: RateLimitPluginOptions = {
   timeWindow: 5000,
   cache: 10000,
   allowList: ['127.0.0.1'],
-  redis: new ioredis({ host: '127.0.0.1' }),
+  redis: process.env.SKIP_REDIS ? undefined : new ioredis({ host: '127.0.0.1' }),
```
Comment (mcollina) [inline test/types/types.test.ts:38]: Why this change?

File: test/route-rate-limit.test.js [hunk 4]
```diff
@@ -1199,7 +1210,7 @@ test('When continue exceeding is on (Local)', async t => {
 test('When continue exceeding is on (Redis)', async t => {
   const fastify = Fastify()
 
-  const redis = await new Redis({ host: REDIS_HOST })
+  const redis = isSkipRedis ? undefined : await new Redis({ host: REDIS_HOST })
```
Comment (mcollina) [inline test/route-rate-limit.test.js:1213]: test have a `skip` option, use that instead

File: test/route-rate-limit.test.js [hunk 5]
```diff
@@ -252,6 +258,11 @@ test('With redis store', async t => {
 })
 
 test('Skip on redis error', async t => {
+  if (isSkipRedis) {
+    t.pass('Redis is not available')
+    return
+  }
```
Comment (mcollina) [inline test/route-rate-limit.test.js:264]: test have a `skip` option, use that instead

**Review by mcollina — COMMENTED** [review]
Why all the changes to redis in the tests?

**Comment by leomp12** [comment]
@mcollina because I got a little lazy to run a docker with redis 🙃
Test were failing with tons of errors without redis connection, so I've done some edits to handle an optional `SKIP_REDIS` env and then I run tests with `SKIP_REDIS=true npm run test`.

But I didn't change anything related to Redis in the lib or in the tests, just skipping some tests if this env is defined.

**Comment by leomp12** [comment]
But I can see here now that the GH Actions workflows are using node 10, and then they fail in the conditional chain I inserted, I'll fix it ..

**Comment by leomp12** [comment]
I can also just revert those changes with `SKIP_REDIS` if you prefer

File: test/types/types.test.ts (hunk 3 above)
Comment (leomp12) [reply test/types/types.test.ts]: unnecessary, reverted..

File: test/global-rate-limit.test.js [hunk 6]
```diff
@@ -241,6 +242,11 @@ test('With function allowList', async t => {
 })
 
 test('With redis store', async t => {
+  if (isSkipRedis) {
+    t.pass('Redis is not available')
+    return
```
[the remark this answers is not in the document]
Comment (leomp12) [reply test/global-rate-limit.test.js]: true, better..

[the remark this answers is not in the document]
Comment (leomp12) [reply test/global-rate-limit.test.js]:
```console
Suites:   ​4 passed​, ​4 of 4 completed​
Asserts:  ​​​2509 passed​, ​7 skip​, ​of 2516​
```

done..

**Comment by leomp12** [comment]
Just to mention, the lines deleted in the `test/global-rate-limit.test.js` file diff are because I removed this duplicate test here https://github.com/fastify/fastify-rate-limit/blob/master/test/global-rate-limit.test.js#L537-L583

File: test/global-rate-limit.test.js [hunk 7]
```diff
@@ -10,6 +10,7 @@ const noop = () => { }
 const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms))
 
 const REDIS_HOST = '127.0.0.1'
+const isSkipRedis = Boolean(process.env.SKIP_REDIS)
```
Comment (Eomm) [inline test/global-rate-limit.test.js:13]: this code is not needed, please remove it and commit using `git commit -n -m 'remove skip redis'` to skip the git hook

Comment (leomp12) [reply test/global-rate-limit.test.js]: okay, done

**Review by mcollina — APPROVED** [review]
lgtm, this looks good

File: README.md [hunk 8]
```diff
@@ -193,6 +194,24 @@ fastify.register(require('fastify-rate-limit'), {
...
+  keyGenerator: function(req) {
+    return return req.userId || req.ip
+  }
+})
+
+fastify.decorateRequest('userId', '')
+fastify.addHook('preHandler', async req => {
+  const { userId } = req.query
+  if (userId) {
+    req.userId = userId
```
Comment (darkgl0w) [inline README.md:210]:
```suggestion
  keyGenerator: function (request) {
    return request.userId || request.ip
  }
})

fastify.decorateRequest('userId', '')
fastify.addHook('preHandler', async function (request) {
  const { userId } = request.query
  if (userId) {
    request.userId = userId
```

As a "convention" we use `request` over `req` and `reply` over `res` across all fastify org repositories.

File: test/global-rate-limit.test.js [hunk 9]
```diff
@@ -1034,3 +1009,49 @@ test('When continue exceeding is on (Redis)', async t => {
...
+
+  fastify.decorateRequest('userId', '')
+  fastify.addHook('preHandler', async req => {
+    const { userId } = req.query
+    if (userId) {
+      req.userId = userId
+    }
+  })
+
+  fastify.get('/', async (req, reply) => 'pão de queijo com café!')
```
Comment (darkgl0w) [inline test/global-rate-limit.test.js:1033]:
```suggestion
  fastify.get('/', async (req, reply) => 'fastify is awesome !')
```

File: test/route-rate-limit.test.js [hunk 10]
```diff
@@ -1266,3 +1265,55 @@ test('should consider routes allow list', t => {
...
+      rateLimit: {
+        max: 1,
+        timeWindow: 10000,
+        hook: 'preValidation',
+        keyGenerator (req) {
+          return req.userId || req.ip
+        }
+      }
+    }
+  }, async (req, reply) => 'pão de queijo com café!')
```
Comment (darkgl0w) [inline test/route-rate-limit.test.js:1287]:
```suggestion
  }, async (req, reply) => 'fastify is awesome !')
```

**Review by darkgl0w — CHANGES_REQUESTED** [review]
A few nitpicks.
Typescript definition unit tests are missing.

Can you please add a unit test inside `test/types/types.test.ts ?

File: README.md (hunk 8 above)
Comment (leomp12) [reply README.md:210]: Personally I also use `request`, but in the examples here `req` was more frequent, so I followed...

File: test/global-rate-limit.test.js (hunk 9 above)
Comment (leomp12) [reply test/global-rate-limit.test.js:1033]: Hahaha 😄

File: test/types/types.test.ts [hunk 11]
```diff
@@ -61,21 +62,24 @@ const options2 = {
...
+  hook: 'preParsing' as RateLimitHook
 }
 
 const options3 = {
   global: true,
   max: (req: FastifyRequest<RequestGenericInterface>, key: string) => 42,
   timeWindow: 5000,
-  store: CustomStore
+  store: CustomStore,
+  hook: 'preValidation' as RateLimitHook
```
Comment (darkgl0w) [inline test/types/types.test.ts:74]: Same here.

File: test/types/types.test.ts [hunk 12]
```diff
@@ -61,21 +62,24 @@ const options2 = {
...
 }
 
 const options4 = {
   global: true,
   max: (req: FastifyRequest<RequestGenericInterface>, key: string) => Promise.resolve(42),
   timeWindow: 5000,
-  store: CustomStore
+  store: CustomStore,
+  hook: 'preHandler' as RateLimitHook
```
Comment (darkgl0w) [inline test/types/types.test.ts:82]: Same here.

File: test/types/types.test.ts [hunk 13]
```diff
@@ -61,21 +62,24 @@ const options2 = {
   global: true,
   max: (req: FastifyRequest<RequestGenericInterface>, key: string) => 42,
   allowList: (req: FastifyRequest<RequestGenericInterface>, key: string) => false,
-  timeWindow: 5000
+  timeWindow: 5000,
+  hook: 'preParsing' as RateLimitHook
```
[the remark this answers is not in the document]
Comment (leomp12) [reply test/types/types.test.ts:66]: Why would `expectAssignable` be useful here?
Since I'm enumerating valid hook options, directly using literal string fires:

```console
Type 'string' is not assignable to type 'RateLimitHook | undefined'.
```

To avoid the workaround with ` as RateLimitHook` (😬) I can explicitly set `let hook: RateLimitHook` before options objects.

[the remark this answers is not in the document]
Comment (darkgl0w) [reply test/types/types.test.ts]: Oups my bad I forgot to add `const option2: RateLimitPluginOptions = ...` and same goes for option3 and option4.

With this it will even trigger a typescript error if the value of hook is not valid.

[the remark this answers is not in the document]
Comment (leomp12) [reply test/types/types.test.ts]: ooh yes! truth 🙃

**Review by Eomm — APPROVED** [review]
